### ISSUE TITLE
Upgradeable Contracts

### ISSUE THREAD

**Issue by jmank88** [gochain-io/explorer#274]
The web3 upgradeable interfaces ([`package assets`](https://godoc.org/github.com/gochain-io/web3/assets)`UpgradeableProxyABI`, `OwnerUpgradeableProxyABI`) should be automatically detected (related #225, #255).

**Comment by jmank88** [comment]
It would be useful if the result of `target()` was automatically displayed as a link to the target contract page.

**Comment by jmank88** [comment]
Interfaces implemented by target contracts should be automatically detected like normal, but we should consider *blocking* interaction from explorer when we know that a contract is actually the target of an upgradable proxy, since the target state is not valid. Additionally, we should *enable* interaction from the proxy, even though it doesn't directly implement the interfaces.

@benbjohnson Am I getting this right?

**Comment by benbjohnson** [comment]
Yes, if we can block the target contract interactions that would be good but we don't currently have a flag indicating that it is a target.

The ABI is totally opaque to the proxy since it just uses the fallback function to pass along to the target contract. Functions in the ABI are assigned a 4-byte hash so if that hash doesn't match any functions in the proxy contract (e.g. `upgrade()`) then it sends it along as the argument to the target via `delegatecall`.

**Comment by jmank88** [comment]
> Yes, if we can block the target contract interactions that would be good but we don't currently have a flag indicating that it is a target.

Yeah I was having second thoughts about this. Explorer can figure out if a proxy targets it - maybe then it  should have a warning like: "This contract is targeted by upgradable proxy 0x1234"
